PXC-4559 Use PXB's reduced lock feature in SST scripts

aybek opened the pull request:
PXB 8.4.0-1 introduces support for the --lock-ddl=REDUCED option. This commit updates PXC to leverage this feature:

1. Use --lock-ddl=REDUCED by default when supported by the installed PXB version.
2. Read and apply the --lock-ddl option from the [xtrabackup] section of my.cnf, if specified.

satya-bodapati commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1678,7 +1707,7 @@ function initialize_pxb_commands()
                 --target-dir=\${DATA} 2>&1 | logger -p daemon.err -t ${ssystag}innobackupex-move "
             INNOBACKUP="${pxb_bin_path} --defaults-file=${WSREP_SST_OPT_CONF} \
                 --defaults-group=mysqld${WSREP_SST_OPT_CONF_SUFFIX} $disver $iopts \
```
we should provide a way for the user to turn off the feature. Here the cnf file is passed first and then the --lock-ddl parameter.  So even if  there is --lock-ddl=ON in cnf, we will still override because we pass --lock-ddl=ON --lock-ddl=reduced and pxb will use the last passed option value.

aybek replied:
Ok so PXB requires --defaults-file to be the very first argument on the command line:
`[ERROR] [MY-011825] [Xtrabackup] --defaults-file must be specified first on the command line`
However, with the current implementation, we already validate the user-provided arguments, so an override issue should not occur.

satya-bodapati commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1620,6 +1622,33 @@ function initialize_pxb_commands()
     fi
 
+    # Check if xtrabackup supports --lock-ddl=REDUCED
+    if ${pxb_bin_path} --help 2>/dev/null | sed -n '/^ *--lock-ddl\[=name\]/,/^ *--/p' | grep -q 'REDUCED'; then
```
I think you can simply detect based on the --version output. if it has -pro, it will have reduced option enabled.

aybek replied:
thanks

satya-bodapati commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1620,6 +1622,33 @@ function initialize_pxb_commands()
+        # Use --lock-ddl=REDUCED only if xtrabackup supports it 
+        # otherwise use the default --lock-ddl=ON;
+        if [[ "$supports_reduced_lock_ddl" == true ]]; then
+            lock_ddl="--lock-ddl=REDUCED"
```
Can we also add a message to the log, that PXB PRO is detected and lock-ddl=reduced is used for SST?

kamil-holubicki commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1620,6 +1622,26 @@ function initialize_pxb_commands()
+        # Use user-provided option or default to REDUCED
+        lock_ddl="--lock-ddl=${lock_ddl_opt:-REDUCED}"
+        
+        [[ -z "$lock_ddl_opt" ]] && wsrep_log_info "PXB Pro detected; using default --lock-ddl=REDUCED for SST"
```
I think this one will print the message also in the case when we've got PXB-pro, and user provided `lock-ddl=SOMETHING_ELSE`

aybek replied:
thanks , updated

aybek committed:
PXC-4559 Use PXB's reduced lock feature in SST scripts
PXB 8.4.0-1 introduces support for the --lock-ddl=REDUCED option.
This commit updates PXC to leverage this feature:

1. Use --lock-ddl=REDUCED by default when supported by the installed PXB version.
2. Read and apply the --lock-ddl option from the [xtrabackup] section of my.cnf, if specified.

aybek commented:
Created new PR with proper branch name: https://github.com/percona/percona-xtradb-cluster/pull/2122
closing this one
